### ISSUE TITLE
One-pager: 🌍 Upstream for the web

### ISSUE THREAD

**Issue by efstajas** [radicle-dev/radicle-upstream#2853]
Currently, Upstream is an Electron app that runs locally on user's machines — and it's running the local "Upstream proxy" to interact with the user's local Radicle monorepo and to enable participation in the Radicle P2P network.

Meanwhile, the [Radicle Network](https://app.radicle.network/) web app maintained by the alt-clients team enables connecting directly to a Radicle seed node to list and view hosted projects without requiring the user to locally track and replicate a project. At the moment, this web app and Upstream are wholly compatible; projects published to the network via the Rad CLI appear on the web app, and can be tracked with Upstream without issue.

Most recently however, Upstream and the radicle.network web app have started somewhat diverging both in focus and implementation approach, resulting in both building out social code collaboration features such as Patches and Issues with full commenting and status tracking features, but built on different incompatible technical approaches. In practice, this means that as soon as patch listing and -commenting go live on radicle.network, patches created within Upstream will not appear on the web app, and vice-versa.

This one-pager aims to lay out a transition approach for how the upstream team may build a web app based on the existing Upstream code base, allowing projects to be viewed within Upstream on the web directly from a seed node similarly to the radicle.network web app. Going a step further, the aim would be to eventually allow the user to run only the upstream proxy locally, which the Upstream web app can then connect to, unlocking the full Upstream collaboration experience from within a browser.

## 🔎 Background & Hypothesis

Note: The below points are as of now fully hypothetical, and should be validated with (potential) users before informing decisions.

### 👀 Open source projects demand their projects being visible online with minimal friction

Open source projects using their public GitHub repo as a de-facto homepage is a frequent sight, and a vital tool for public discovery and education. In order for a project to consider an alternative code hosting solution, said solution needs to make projects publicly visible, and ensure they can be indexed on search engines.

### 👎 Users are hesitant to install software locally

With Upstream as a desktop-only client, users are required to install a piece of software locally upfront before being able to view an Upstream project, as well as open an issue or view open patches. This constitutes a high barrier of entry, and likely leads to a bounce early in the adoption journey — a user's intent of, say, opening an issue for a bug they ran into with a particular project may not be strong enough to justify an upfront local installation. For this reason, we should ensure as much functionality as possible can be accessed without requiring any local installation.

### 👎 Jumping between different projects and UIs as part of a typical adoption journey is messy and elevates bounce rates

Our goal is to get users to use Upstream — and to do that, we need to ensure that the initial discovery and onboarding process is as smooth as can be. At the moment, users looking for a project hosted on Radicle are likely to initially discover the project on the radicle.network web app, which offers no entry point to installing upstream and interacting with the project (by, say, opening a patch). Even if someone were to separately discover the project and the Upstream client, they would be exposed to a vastly different user experience within upstream compared to the web app, employing divergent UX patterns, wording and concepts. And with patches and issues launching on radicle.network, a project that decides to use Upstream patches would see an empty patch list on the radicle.network web app, further adding to the confusion.

### 👍 A web-first approach results in a smoother adoption journey

If the user were to use the same exact UI to first discover and later install Radicle and interact with a project, their adoption journey would be more straightforward, no longer require jumping between different tools, and consistent in user experience and UI patterns.

## 💡 Solution proposal: Bringing Upstream to the web

Given all the above, let's consider how we could build a version of upstream that runs on the web, with full public discoverability, while not losing any of its abilities of interacting with the local filesystem, user identity, and writing to the Radicle network.

#### 1. Decoupling Upstream components from their state

Currently, various views that would be relevant for the web are rather tightly coupled with their state — for example, the `Source` component directly depends on the `ProjectScreen` view state that it is nested under. Since the project screen would need to be accessible pre local installation, it needs to be able to receive data fetched directly from a seed node. This would mean slight differences in data model, and require a different approach to fetching data (especially with server-side rendering on the web). Thus, our first step should be slowly decoupling various components of the `ProjectScreen` into a pure representational version that intakes all state via props, and then wrapping said components in Upstream to provide state from the same places it's already coming from today.

#### 2. Launch a web version of the project page

With the `ProjectScreen` components being purely representational, we can create a separate application in a subdirectory of this repo, perhaps even built on SvelteKit for out-of-the-box server-side rendering. We can depend on the exact same representational components as the upstream app does today in order to build a web version of the project page that allows viewing a project hosted on a seed node directly, very similar to the [radicle.network project page](https://app.radicle.network/seeds/maple.radicle.garden/rad:git:hnrk81fjxemf69dqc5knbmtc6frdi1n3rbh9o/tree/65e4ba4eee4fbddae19afaac8dae54725677d90a), except using Upstream's project view design and components. The Patches (and depending on progress, Issues) tabs may already be visible here, but initially just prompt the user to download and install Upstream locally in order to see patches & issues.

👍 At this point, we already allow discovery of projects online, with a clear entry point to downloading Upstream locally. Throughout this journey the user will be interacting with the same application, adding visual and UX consistency.

👎 Users are still required to install something locally in order to even view patches and issues — resulting in a likely early bounce.

#### 3. List Upstream patches & issues on the web version

Next, we can find a way to list Upstream patches & issues, including comments and statuses, directly on this web "preview" of a project. For this, we would need to contribute the necessary endpoints to the node HTTP API, or run our own node (?). Once this is done, we can display full patches and issues on the web, and prompt the user to install Upstream locally as soon as they want to write to the network themselves — for example by opening a new issue, or leaving a comment on a patch. This moves the local installation one step later, hypothetically significantly reducing the bounce rate, given that users are able to form a stronger intent before being presented with the necessity of a local installation.

![image](https://user-images.githubusercontent.com/1018218/172583220-3712a467-0b7f-4fcc-8a73-30a80e3e3142.png)

👍 At this point, an entire Upstream project can be fully previewed on the web, and the user can already view issues alongside code before any local installation.

👎 Users need to download and the entire Upstream application locally, resulting in a jump between two environments as part of a typical first use journey.

#### 4. Bring discovery to the Upstream web app

Next up, we may add basic discovery functionality to the Upstream web app, similar to [radicle.network's project listing page](https://app.radicle.network/seeds/maple.radicle.garden). This would allow users to search and explore projects hosted on the network, further strengthening discoverability of individual projects.

#### 5. Bring the full Upstream experience to the web

The last step is to "demote" the local installation from the full Upstream application to a background application that simply locally hosts the Upstream proxy (or something similar). This proxy would expose a local web server that the web app can talk to (secured by CORS — or later some kind of authentication system, for example SIWE or a browser extension using [native messaging](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging)). Instead of downloading Upstream and using it locally, users would be able to run only this background application. The web app could detect said local proxy, and switch over to a mode that unlocks full write functionality, including commenting on a patch and creating an issue. Effectively, as soon as the proxy is installed locally, the app unlocks the full Upstream experience as we know it, except all _while staying in the same browser tab that the user initially landed on_.

![image](https://user-images.githubusercontent.com/1018218/172583135-cf2ed53a-cea6-4938-9687-27d261b33d62.png)

How exactly to do this technically is to be discussed later-on, but one way may even be to simply convert the existing Upstream electron app into a lean [menu bar app](https://github.com/maxogden/menubar) that automatically sets itself to run on startup. It could look something like this:

![image](https://user-images.githubusercontent.com/1018218/172588689-3be4de59-4276-440e-aa6f-5e4ab808f1b0.png)

👍 The entire adoption user journey is one straight line, with a smooth onboarding process that does not require jumping between environments any longer.

👍 Users can easily share their Upstream projects and even specific comments or issues by simply copying the link they're on.

👍 We have full control over the frontend version, and can prompt users to upgrade their local proxy-like application if necessary.

**Comment by rudolfs** [comment]
Nice work @efstajas! Really like the one-pager format.

I agree that a web-first Upstream would lower the barrier of entry significantly.

What would be the first user problem we want to solve once we would have a web native Upstream? Is project source browsing an important enough problem to solve first? I assume it's to do with discoverability, but would it not make more sense to directly focus on code collaboration and just build anything that's needed to contribute code, i.e. the patches workflow?

At the moment we have control of the whole distribution because it's packaged together, this ensures the front-end, back-end and CLI are all compatible. How do we make sure that the locally running Upstream and the web Upstream are compatible, especially when people start self hosting and there are multiple different versions of nodes in the wild?

I'm wondering whether having SIWE as the only way to use Upstream would mean potentially excluding people from the p2p crowd?

**Comment by efstajas** [comment]
Thanks @rudolfs! Good points.

> What would be the first user problem we want to solve once we would have a web native Upstream? Is project source browsing an important enough problem to solve first?

I am operating under the assumption that we would want to bring Upstream functionality as-is to the web initially and iterate from there — meaning that at least for a while, general code hosting would still be the foundational feature everything else is built on top. And that coupled with hypothesis #1 ("Open source projects demand their projects being visible online with minimal friction") would make source browsing a natural first step to bring to the web. I'd argue that having Upstream's source browsing functionality on the web makes it a whole lot more valuable, given that currently Upstream's project view is only accessible locally — where users already have a text editor to browse their code more conveniently anyway.

> At the moment we have control of the whole distribution because it's packaged together, this ensures the front-end, back-end and CLI are all compatible. How do we make sure that the locally running Upstream and the web Upstream are compatible, especially when people start self hosting and there are multiple different versions of nodes in the wild?

Regarding the web upstream and locally running proxy: with the web app being hosted on our servers, we can freely upgrade it and no longer rely on the user updating their UI manually. I would propose that the locally running proxy exposes a version endpoint, which the Upstream web app can query. In the situation where an update of the locally running proxy is required for a new feature, we can prompt the user to update their local proxy upon attempting to access said new feature — and in case of a breaking change, we may even block the UI entirely until the user updates the local version.

Regarding seed nodes: I suppose this is an unsolved problem now already, but the impact is limited given the API is relatively stable and versioned. I'm not sure if seed nodes already have some kind of version endpoint, but if they do, a similar pattern to the one could be applied — selectively disabling unsupported functionality in case a seed node is running an outdated version.

> I'm wondering whether having SIWE as the only way to use Upstream would mean potentially excluding people from the p2p crowd?

Unless I'm missing something, just setting a CORS policy on the proxy to only allow requests from our web app (or localhost) is enough to secure things initially, and no SIWE or any other authentication mechanism would be required. Only if we ever want to enable third party apps connecting to the local Upstream proxy, we need some kind of authentication mechanism. With SIWE, users could configure their Ethereum address on the local proxy, and third party apps could connect to it with a signed message. If we don't want to rely on the user having an Ethereum wallet for this, I suppose even a simple passkey-based authentiation for new connections would suffice. Alternatively, a browser extension communicating to the proxy via native messaging (while not exposing an http API on localhost) could allow users to selectively grant access to the local proxy to specific sites.

**Comment by geigerzaehler** [comment]
Thanks for the great write-up @efstajas

I have a suggestion around the ordering of the steps in the approach which also considers the fact that Upstream is still in an early product stage.

My argument is that step 2 and 3 do too little to improve on the problem and may make it event worse in some aspects.

After step 3 we users can inspect projects on the web with a consistent look and feel. But we still need `radicle.network` for discovery and we now have three apps (not counting the CLI) that the user can interact with. I doubt that a simple project inspector that shares Upstreams look and feel is an improvement to the users’ onboarding journey. In addition, most of the risk is in the last two steps which we should try to address as early as possible.

Instead I propose to add discovery to the desktop app first. This does not move the installation step to a later point in the user journey and we can’t leverage the accessibility of the web but the user journey now happens in just one app.

The assumption here is that there is no difference in effort between implementing discovery in Electron first and then porting the App to the web or the other way around.

Another point to consider is that the biggest unknown in the development of this project (in terms of expected behavior and implementation) is how we present the mixture of local and seed node data. We should try to release a proof of concept for this as soon as possible. However, if we can only release this together with the elimination of the desktop app we loose an opportunity to get feedback early. (Keep in mind that we can only eliminate the desktop app after we’ve developed an alternative distribution method that can run the background process.) I’d rather start working on this prollem in the desktop app which is possible after integrating discovery.
